File: apps/embedding-service/vectors.py
import math
from collections.abc import Iterable
from typing import Any


def normalize_vector(vector: list[float]) -> list[float]:
    norm = math.sqrt(sum(value * value for value in vector))
    if norm == 0:
        raise ValueError("model returned a zero vector")
    return [value / norm for value in vector]


def _float_vector(value: Any) -> list[float]:
    if not isinstance(value, list) or not all(isinstance(item, int | float) for item in value):
        raise ValueError("model returned an invalid embedding vector")
    return [float(item) for item in value]


def extract_embedding_vectors(result: Any) -> list[list[float]]:
    if not isinstance(result, dict):
        raise ValueError("model returned an invalid embedding response")
    data = result.get("data")
    if not isinstance(data, list):
        raise ValueError("model returned an invalid embedding response")

    vectors: list[list[float]] = []
    for item in data:
        if not isinstance(item, dict):
            raise ValueError("model returned an invalid embedding response")
        vectors.append(_float_vector(item.get("embedding")))
    return vectors


def normalize_vectors(vectors: Iterable[list[float]]) -> list[list[float]]:
    return [normalize_vector(vector) for vector in vectors]

File: apps/embedding-service/src/vectors.ts
export const normalizeVector = (vector: number[]): number[] => {
  const norm = Math.sqrt(vector.reduce((sum, value) => sum + value * value, 0));
  if (norm === 0) {
    throw new Error("model returned a zero vector");
  }
  return vector.map((value) => value / norm);
};

const floatVector = (value: unknown): number[] => {
  if (
    !Array.isArray(value) ||
    !value.every((item) => typeof item === "number" && Number.isFinite(item))
  ) {
    throw new Error("model returned an invalid embedding vector");
  }
  return value.map(Number);
};

export const extractEmbeddingVectors = (result: unknown): number[][] => {
  if (typeof result !== "object" || result === null) {
    throw new Error("model returned an invalid embedding response");
  }
  const data = (result as { data?: unknown }).data;
  if (!Array.isArray(data)) {
    throw new Error("model returned an invalid embedding response");
  }
  return data.map((item) => {
    if (typeof item !== "object" || item === null) {
      throw new Error("model returned an invalid embedding response");
    }
    return floatVector((item as { embedding?: unknown }).embedding);
  });
};

export const normalizeVectors = (vectors: number[][]): number[][] =>
  vectors.map(normalizeVector);
